### PR TITLE
LUCENE-10229: return -1 for unknown offsets in ExtendedIntervalsSource. Modify highlighting to work properly with or without offsets (depending on their availability).

### PR DESCRIPTION
**Pull request by dweiss**

### CODE REVIEW

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/ConjunctionMatchesIterator.java [hunk 1]
```diff
@@ -60,7 +60,11 @@ public int endPosition() {
   public int startOffset() throws IOException {
     int start = Integer.MAX_VALUE;
     for (MatchesIterator s : subs) {
```
Comment (dweiss) [inline lucene/queries/src/java/org/apache/lucene/queries/intervals/ConjunctionMatchesIterator.java:62]: If any source offset is unknown, propagate it up.

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/ExtendedIntervalsSource.java [hunk 2]
```diff
@@ -53,9 +55,60 @@ public IntervalMatchesIterator matches(String field, LeafReaderContext ctx, int
...
+          @Override
+          public int startPosition() {
+            return delegate.startPosition();
+          }
+
+          @Override
+          public int endPosition() {
+            return delegate.endPosition();
+          }
+
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/ExtendedIntervalsSource.java:87]: I think we should always return -1 here? After all, we will always be extending things in at least one direction, so we will never be able to return correct offsets for the whole interval.

**Review by romseygeek — APPROVED** [review]
This is a very nice solution! I left one nit but other than that it looks good to me.

**Comment by dweiss** [comment]
> I think we should always return -1 here? After all, we will always be extending things in at least one direction, so we will never be able to return correct offsets for the whole interval.

I was thinking about it too. I think sometimes you could "correct" the unknown entire expressions you sometimes could compute the correct result, assuming position/offset pairs are correspondingly monotonic (where min/max offsets are computed from sub-intervals). I think it is (now) required that position(a).startOffset <= position(b).startOffset for a < b, right?

**Comment by dweiss** [comment]
I did take another look and even though the start offsets are monotonic, the end offsets don't have to be (with respect to positions). Let's leave it simple for now. I've changed the returned offset from both methods to -1 for consistency.

**Pull request merged**